app/scripts/lib/transaction/swaps.ts

```diff
@@ -0,0 +1,122 @@
+import { pickBy, merge } from 'lodash';
+import {
+  TransactionStatus,
+  TransactionType,
+  TransactionMeta,
+} from '../../../../shared/constants/transaction';
+import MetamaskController from '../../metamask-controller';
+
+type SwapOptions = {
+  // Whether this transaction required an approval transaction
+  hasApproveTx: boolean;
+  // Additional metadata to store for the transaction
+  meta: any;
+};
+
+type SwapApprovalTransaction = {
+  sourceTokenSymbol: string;
+  destinationTokenSymbol: string;
+  type: TransactionType;
+  destinationTokenDecimals: string;
+  destinationTokenAddress: string;
+  swapMetaData: string;
+  swapTokenValue: string;
+  estimatedBaseFee: string;
+  approvalTxId: string;
+};
+
+/**
+ * Creates a new transaction with swap-specific metadata
+ *
+ * @param this - MetamaskController
+ * @param swapOptions - Swap-specific options
+ * @param transactionType - The type of transaction
+ * @param transactionMeta - The transaction metadata
+ * @returns The updated transaction metadata
+ */
+export async function createSwapsTransaction(
+  this: MetamaskController,
```

Comment:
I fear this is way too large a type to include in this callback, plus maybe a little confusing to call an argument `this`.

Should we instead create an additional type such as `UpdateSwapsTransactionRequest` that includes callbacks like `setApproveTxId`, `updateTransactionSwapProperties`, and `setTradeTxId`?

Similar to what we did with metrics.

Reply:
This is removed now. 
But just wanted to mention why I did bind entire object. While binding those functions to callback, it was not possible to read the `txController.X` function because `txController` was not defined yet in the context, same reason for swaps callbacks. 

I didn't encounter this issue with `getTransactionEventRequest` because it's safe to call it right after initialisation of `TxController`